[Inductor] Add note that stride_vars result may be inaccurate

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89877
* __->__ #90184

Strides are are determined by substituting 1 and 0 for different indices, which
will fail for any expression that doesn't match the expected stride calculation.
So, lets add a note to make this clear.

peterbell10 committed:
[Inductor] More robust stride and offset extraction from index expressions

Currently the stride and offset are determined by substituting 1 and 0 for
different indices, which will fail for any expression that doesn't match the
expected stride calculation. Instead, this uses `sympy.match` and returns `None`
for any variables used in non-standard index calculation, e.g. `torch.roll`
which uses `ModularIndexing`.

[ghstack-poisoned]

peterbell10 committed:
Update on "[Inductor] More robust stride and offset extraction from index expressions"


Currently the stride and offset are determined by substituting 1 and 0 for
different indices, which will fail for any expression that doesn't match the
expected stride calculation. Instead, this uses `sympy.match` and returns `None`
for any variables used in non-standard index calculation, e.g. `torch.roll`
which uses `ModularIndexing`.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge -f "CI is green"

ngimel commented:
This catastrophically regresses performance (apparently for all indexing expressions that have ModularIndexing), e.g. hg_Bert 1.20x -> 1.029x, timm_vision_transformer 1.49x->1.23x. 
```
python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --float16 --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt) --only hf_Bert 
```
```
TORCHINDUCTOR_SHAPE_PADDING=1 python benchmarks/dynamo/torchbench.py  --performance  --backend inductor --float16 --training --batch-size-file $(realpath benchmarks/dynamo/torchbench_models_list.txt) --only timm_vision_transformer 
```
I suggest we revert it.

soumith commented:
yea let's revert it

ngimel commented:
@pytorchbot revert -m "catastrophically regresses performance"

ngimel commented:
@pytorchbot revert -m "catastrophically regresses performance" -c nosignal

peterbell10 committed:
Update on "[Inductor] More robust stride and offset extraction from index expressions"


Currently the stride and offset are determined by substituting 1 and 0 for
different indices, which will fail for any expression that doesn't match the
expected stride calculation. Instead, this uses `sympy.match` and returns `None`
for any variables used in non-standard index calculation, e.g. `torch.roll`
which uses `ModularIndexing`.

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/sizevars.py:
```diff
@@ -384,52 +389,86 @@ def wrapper(*args, **kwargs):
         result = []
-        for s in self.stride_vars(index, vars):
+        for i, s in enumerate(self.maybe_stride_vars(index, vars)):
+            if s is None:
+                # If the dimension is not truly strided, we approximate a
+                # stride by taking the difference of the 1st and 0th element in
+                # that dimension.
+                zero_index = {v: sympy.Integer(0) for v in vars if v != 0}
+                first_index = dict(zero_index)
+                first_index[vars[i]] = sympy.Integer(1)
+                s = sympy_subs(index, dim_only) - sympy_subs(index, zero_index)
```
This fixes the regression for me locally. `decide_layout()` was using the stride_hint of zero to always put the `ModularIndex` dimensions first. In that context I think using the old method is okay since you shouldn't be relying on `stride_hint` for correctness anyway.

peterbell10 committed:
Update on "[Inductor] Add note that stride_vars result may be inaccurate"


Strides are are determined by substituting 1 and 0 for different indices, which
will fail for any expression that doesn't match the expected stride calculation.
So, lets add a note to make this clear.

[ghstack-poisoned]

peterbell10 commented:
Actually, after a second look I think `stride_hint` is the main user anyway so adding an escape hatch for it is over-complicating things. I've changed this PR to just add a note so nobody else gets tripped up by this in future.

peterbell10 commented:
@pytorchbot merge

peterbell10 commented:
@pytorchbot merge
